tests/functional/tests/security/test_metadata_corruption.py

```diff
@@ -0,0 +1,334 @@
+    )
+
+    corrupted_byte = get_random_target_in_segment(cache, target_segment)
+
+    core = Core(core_volume)
+    cache.add_core(core)
+
+    core_exp_volume = CoreVolume(core)
+    queue = cache.get_default_queue()
+
+    r = (
+        Rio()
+        .target(core_exp_volume)
+        .njobs(1)
+        .readwrite(ReadWrite.WRITE)
+        .size(Size.from_MiB(1))
+        .qd(1)
+        .run([queue])
+    )
+
+    if recovery:
+        cache.save()
+        cache.device.offline()
+
+    exc = False
+    try:
+        cache.stop()
+    except OcfError:
+        exc = True
+
+    cache_volume.online()
+
+    assert recovery == exc, "Stopping with device offlined should raise an exception"
+
+    corrupt_byte(cache_volume.data, corrupted_byte)
+
+    with expectation:
+        cache = Cache.load_from_device(cache_volume)
+
+
+@pytest.mark.security
+@pytest.mark.parametrize("cache_line_size", CacheLineSize)
+@pytest.mark.parametrize("cache_mode", CacheMode)
+@pytest.mark.parametrize("target_segment", CacheMetadataSegment)
+def test_metadata_corruption_standby_activate(
+    pyocf_2_ctx, cache_line_size, cache_mode, target_segment
+):
+    primary_ctx, secondary_ctx = pyocf_2_ctx
+
+    primary_cache_volume = RamVolume(Size.from_MiB(60))
+    secondary_cache_volume = RamVolume(Size.from_MiB(60))
+
+    core_volume = RamVolume(Size.from_MiB(1))
+
+    secondary_cache = Cache(
+        owner=secondary_ctx,
+        cache_mode=cache_mode,
+        cache_line_size=cache_line_size,
+        name="secondary",
+    )
+    secondary_cache.start_cache()
+    secondary_cache.standby_attach(secondary_cache_volume)
+
+    corrupted_byte = get_random_target_in_segment(secondary_cache, target_segment)
+
+    secondary_cache_exp_obj = CacheVolume(secondary_cache)
+    primary_cache_replicated_volume = ReplicatedVolume(
+        primary_cache_volume, secondary_cache_exp_obj
+    )
+
+    primary_cache = Cache.start_on_device(
+        primary_cache_replicated_volume,
+        owner=primary_ctx,
+        cache_mode=cache_mode,
+        cache_line_size=cache_line_size,
+        name="primary",
+    )
+    core = Core(core_volume)
+    primary_cache.add_core(core)
+
+    core_exp_volume = CoreVolume(core)
+    queue = primary_cache.get_default_queue()
+
+    r = (
+        Rio()
+        .target(core_exp_volume)
+        .njobs(1)
+        .readwrite(ReadWrite.WRITE)
+        .size(Size.from_MiB(1))
+        .qd(1)
+        .run([queue])
+    )
+
+    primary_cache.save()
+    primary_cache.device.offline()
+
+    exc = False
+    try:
+        primary_cache.stop()
+    except OcfError:
+        exc = True
+
+    primary_cache_replicated_volume.online()
+
+    secondary_cache.standby_detach()
+
+    assert exc, "Stopping with device offlined should raise an exception"
+
+    corrupt_byte(secondary_cache_volume.data, corrupted_byte)
+
+    with pytest.raises(OcfError):
+        secondary_cache.standby_activate(secondary_cache_volume, open_cores=False)
```

Comment:
Every metadata segment would cause activate failure after corrupting it? I thought this would have similar error criteria as a regular recovery, e.g. corruption in LRU will be overwritten by rebuild of this section.

Reply:
whoops, nice catch!